File: src/org/blackpanther/math/Sphere.java
package org.blackpanther.math;

import static java.lang.Math.sqrt;

public class Sphere implements Shape {

    public Point3D getPositiveSurfacePoint(Point3D incompletePoint) {
        final float zComponent =
                (float) sqrt(
                        1f -
                                incompletePoint.getX() * incompletePoint.getX() -
                                incompletePoint.getY() * incompletePoint.getY()
                );
        return new Point3D(
                incompletePoint.getX(),
                incompletePoint.getY(),
                zComponent
        );
    }

    public Point3D getNegativeSurfacePoint(Point3D incompletePoint) {
        final float zComponent =
                (float) -sqrt(
                        1f -
                                incompletePoint.getX() * incompletePoint.getX() -
                                incompletePoint.getY() * incompletePoint.getY()
                );
        return new Point3D(
                incompletePoint.getX(),
                incompletePoint.getY(),
                zComponent
        );
    }

    @Override
    public Shape transform(Transformation trans) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public boolean contains(Point3D point) {
        return sqrt(point.getX() * point.getX() + point.getY() * point.getY() + point.getZ() * point.getZ()) <= 1f;
    }

    @Override
    public String toString() {
        return "Sphere";
    }

}

File: 3d-library/src/model/Sphere.scala
package org.blackpanther.three
package model

import java.awt.geom.Point2D
import scala.math.{sqrt}

/**
 * @author MACHIZAUD Andréa
 * @version 11/5/11
 */

final class Sphere(
  val radius : Float
)extends Shape {

  def transform(trans: Transformation) = this

  override val toString = "Sphere[radius= %.4f]" format radius

  def surfacePointFrom(pt : Point2D) : (ShapeReferential, ShapeReferential) = {
    val height = sqrt(
      1.0 -
      pt.getX * pt.getX -
      pt.getY * pt.getY
    ).toFloat
    (
      new Point3D(pt.getX.toFloat, pt.getY.toFloat,  height) with ShapeReferential,
      new Point3D(pt.getX.toFloat, pt.getY.toFloat, -height) with ShapeReferential
    )
  }

  def contains(pt : ShapeReferential) : Boolean =
    sqrt(pt.x * pt.x + pt.y * pt.y + pt.z * pt.z).toFloat <= radius

  override lazy val hashCode = Sphere.getID() + radius.##
}

object Sphere {

  private var counter = 0

  @inline private[Sphere] def getID() = {
    counter += 1
    counter
  }

  val DefaultRadius = 1f

  def apply(radius : Float = DefaultRadius) = new Sphere(radius)

}